lib/Dialect/HW/Transforms/HWSpecialize.cpp

```diff
@@ -0,0 +1,332 @@
+//===- HWSpecialize.cpp - hw module specialization ------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This transform performs specialization of parametric hw.module's.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/HW/HWAttributes.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/HW/HWPasses.h"
+#include "circt/Dialect/HW/HWSymCache.h"
+#include "circt/Dialect/SV/SVPasses.h"
+#include "circt/Support/ValueMapper.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace llvm;
+using namespace mlir;
+using namespace circt;
+using namespace hw;
+
+namespace {
+
+using InstanceParameters = llvm::DenseMap<hw::HWModuleOp, ArrayAttr>;
+
+// Generates a module name by composing the name of 'moduleOp' and the set of
+// provided 'parameters'.
+static std::string generateModuleName(SymbolCache &symbolCache,
+                                      hw::HWModuleOp moduleOp,
+                                      ArrayAttr parameters) {
+  assert(parameters.size() != 0);
+  std::string name = moduleOp.getName().str();
+  for (auto param : parameters) {
+    auto paramAttr = param.cast<ParamDeclAttr>();
+    int64_t paramValue = paramAttr.getValue().cast<IntegerAttr>().getInt();
+    name += "_" + paramAttr.getName().str() + "_" + std::to_string(paramValue);
+  }
+
+  // Uniqueness check
+  int uniqueCntr = 0;
+  std::string uniqueName = name;
+  auto *ctx = moduleOp.getContext();
+  while (symbolCache.getDefinition(StringAttr::get(ctx, uniqueName)))
+    uniqueName = name + "_" + std::to_string(uniqueCntr++);
+
+  return uniqueName;
+}
+
+// Returns true if any operand or result of 'op' is parametric.
+static bool isParametricOp(Operation *op) {
+  return llvm::any_of(op->getOperandTypes(), isParametricType) ||
+         llvm::any_of(op->getResultTypes(), isParametricType);
+}
+
+// Narrows 'value' using a comb.extract operation to the width of the
+// hw.array-typed 'array'.
+static FailureOr<Value> narrowValueToArrayWidth(OpBuilder &builder, Value array,
+                                                Value value) {
+  OpBuilder::InsertionGuard g(builder);
+  if (value.isa<BlockArgument>())
+    builder.setInsertionPointToStart(value.getParentBlock());
+  else
+    builder.setInsertionPointAfter(value.getDefiningOp());
```

Comment:
Probably you can use `builder.setInsertionPointAfterValue` to do the same thing.

Reply:
Thank you for catching that! will update.